From 0cbbe4bd6d579d377c477b94b3a33c3dac70b4ea Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:10:12 -0400
Subject: [usb] xhci: clear root port wake on bits if controller isn't wake-up capable

Message-id: <1410887463-20674-541-git-send-email-dzickus@redhat.com>
Patchwork-id: 95503
O-Subject: [RHEL7 PATCH 540/591] xhci: clear root port wake on bits if controller isn't wake-up capable
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit ff8cbf250b448aac35589f6075082c3fcad8a8fe
Author: Lu Baolu <baolu.lu@linux.intel.com>
Date:   Tue Jun 24 17:14:43 2014 +0300

    xhci: clear root port wake on bits if controller isn't wake-up capable

    When xHCI PCI host is suspended, if do_wakeup is false in xhci_pci_suspend,
    xhci_bus_suspend needs to clear all root port wake on bits. Otherwise some Intel
    platforms may get a spurious wakeup, even if PCI PME# is disabled.

    This patch should be back-ported to kernels as old as 2.6.37, that
    contains the commit 9777e3ce907d4cb5a513902a87ecd03b52499569
    "USB: xHCI: bus power management implementation".

    Cc: stable@vger.kernel.org # 2.6.37
    Signed-off-by: Lu Baolu <baolu.lu@linux.intel.com>
    Signed-off-by: Mathias Nyman <mathias.nyman@linux.intel.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/xhci-hub.c b/drivers/usb/host/xhci-hub.c
index 2b998c6..aa79e87 100644
--- a/drivers/usb/host/xhci-hub.c
+++ b/drivers/usb/host/xhci-hub.c
@@ -22,6 +22,7 @@
 
 
 #include <linux/slab.h>
+#include <linux/device.h>
 #include <asm/unaligned.h>
 
 #include "xhci.h"
@@ -1139,7 +1140,9 @@ int xhci_bus_suspend(struct usb_hcd *hcd)
    * including the USB 3.0 roothub, but only if CONFIG_PM_RUNTIME
    * is enabled, so also enable remote wake here.
    */
-  if (hcd->self.root_hub->do_remote_wakeup) {
+  if (hcd->self.root_hub->do_remote_wakeup
+    && device_may_wakeup(hcd->self.controller)) {
+
    if (t1 & PORT_CONNECT) {
     t2 |= PORT_WKOC_E | PORT_WKDISC_E;
     t2 &= ~PORT_WKCONN_E;
-- 
1.7.1